internal/receiver/smartagentreceiver/convert.go

```diff
@@ -0,0 +1,186 @@
+		case pdata.MetricDataTypeIntGauge:
+			err = fillIntDatapoint(datapoint, m.IntGauge().DataPoints())
+		case pdata.MetricDataTypeIntSum:
+			err = fillIntDatapoint(datapoint, m.IntSum().DataPoints())
+		case pdata.MetricDataTypeDoubleGauge:
+			err = fillDoubleDatapoint(datapoint, m.DoubleGauge().DataPoints())
+		case pdata.MetricDataTypeDoubleSum:
+			err = fillDoubleDatapoint(datapoint, m.DoubleSum().DataPoints())
+		}
+
+		if err != nil {
+			numDropped++
+			c.logger.Debug("SignalFx datapoint datum conversion error",
+				zap.Error(err),
+				zap.String("metric", datapoint.Metric))
+			continue
+		}
+
+		i++
+	}
+
+	metrics.Resize(i)
+
+	return md, numDropped
+
+}
+func setDataType(datapoint *sfx.Datapoint, m pdata.Metric) error {
+	sfxMetricType := datapoint.MetricType
+	if sfxMetricType == sfx.Timestamp {
+		return errUnsupportedMetricTypeTimestamp
+	}
+
+	var isFloat bool
+	switch datapoint.Value.(type) {
+	case sfx.IntValue:
+		isFloat = false
+	case sfx.FloatValue:
+		isFloat = true
+	default:
+		return fmt.Errorf("unsupported value type %T: %v", datapoint.Value, datapoint.Value)
+	}
+
+	switch sfxMetricType {
+	case sfx.Gauge, sfx.Enum, sfx.Rate:
+		if isFloat {
+			m.SetDataType(pdata.MetricDataTypeDoubleGauge)
+			m.DoubleGauge().InitEmpty() // will need to be removed w/ 0.16.0 adoption
+		} else {
+			m.SetDataType(pdata.MetricDataTypeIntGauge)
+			m.IntGauge().InitEmpty() // will need to be removed w/ 0.16.0 adoption
+		}
+	case sfx.Count:
+		if isFloat {
+			m.SetDataType(pdata.MetricDataTypeDoubleSum)
+			m.DoubleSum().InitEmpty() // will need to be removed w/ 0.16.0 adoption
+			m.DoubleSum().SetAggregationTemporality(pdata.AggregationTemporalityDelta)
+			m.DoubleSum().SetIsMonotonic(true)
+		} else {
+			m.SetDataType(pdata.MetricDataTypeIntSum)
+			m.IntSum().InitEmpty() // will need to be removed w/ 0.16.0 adoption
+			m.IntSum().SetAggregationTemporality(pdata.AggregationTemporalityDelta)
+			m.IntSum().SetIsMonotonic(true)
+		}
+	case sfx.Counter:
+		if isFloat {
+			m.SetDataType(pdata.MetricDataTypeDoubleSum)
+			m.DoubleSum().InitEmpty() // will need to be removed w/ 0.16.0 adoption
+			m.DoubleSum().SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+			m.DoubleSum().SetIsMonotonic(true)
+		} else {
+			m.SetDataType(pdata.MetricDataTypeIntSum)
+			m.IntSum().InitEmpty() // will need to be removed w/ 0.16.0 adoption
+			m.IntSum().SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+			m.IntSum().SetIsMonotonic(true)
+		}
+	default:
+		return fmt.Errorf("unsupported metric type %T: %v", sfxMetricType, sfxMetricType)
+	}
+
+	return nil
+}
+
+func fillIntDatapoint(datapoint *sfx.Datapoint, dps pdata.IntDataPointSlice) error {
+	var intValue sfx.IntValue
+	var ok bool
+	if intValue, ok = datapoint.Value.(sfx.IntValue); !ok {
+		return errNoIntValue
+	}
+
+	dps.Resize(1)
+	dp := dps.At(0)
+	dp.SetTimestamp(pdata.TimestampUnixNano(uint64(datapoint.Timestamp.UnixNano())))
```

Comment:
Should we check for `nil` timestamps on datapoint?

Reply:
since `Timestamp` is a `time.Time` and not a pointer I don't believe nil can be assigned?